---
title: "Tidy ASV and metadata tables"
output: html_notebook
---

Jacobo de la Cuesta-Zuluaga. August 2023.

The aim of this notebook is to process perform alpha diversity analyses on EED 16S data

# Libraries
```{r}
library(tidyverse)
library(conflicted)
library(GUniFrac)
library(vegan)
library(rstatix)
library(ggthemes)
```

```{r}
conflict_prefer("filter", "dplyr")
```

# Prompt
This is a function to remind the user to fill the corresponding fields!
```{r}
Check_point = function(Message){
  Variables_set = askYesNo(msg = Message)
  stopifnot("Make sure the fields are filled!" = isTRUE(Variables_set))
}

```


# Paths
```{r}
Check_point("Are the base and out directories filled?")

# TODO make sure the directories are correct
base_dir = "CHANGE ME!"

EED_16S_dir = file.path(base_dir, "dada2_out")
ref_dir = "Q:/NAS/AG_Maier/Amplicon_Sequencing/Reference_files"

EED_16S_dir %>% 
  list.files(full.names = TRUE, pattern = "EED")

# ASV table
EED_asv_raw = file.path(EED_16S_dir, "2023_11_EED_ASV_table.tsv") %>% 
  read_tsv()

# taxonomy table
EED_full_tax_raw = file.path(EED_16S_dir, "2023_11_EED_taxonomy_full.tsv") %>% 
  read_tsv()

# taxonomy table
EED_taxonomy_raw = file.path(EED_16S_dir, "2023_11_EED_taxonomy_table.tsv") %>% 
  read_tsv()

# Phylogeny
EED_tree_raw =  file.path(EED_16S_dir, "2023_11_EED_tree.tre") %>% 
   ape::read.tree()

# Output dir
output_dir =  file.path(base_dir, "clean_tables")

# Output dir
fig_dir =  file.path(base_dir, "figures")
dir.create(fig_dir)
```

# Clean data tables
# Full ASV table
```{r}
# Remove blanks
EED_asv_full = EED_asv_raw %>% 
  filter(!str_detect(Sample, "Blank"))

# Print
EED_asv_full
```

# EED ASV table
```{r}
# Potential and exact matches to EED taxa
EED_taxonomy = EED_taxonomy_raw %>% 
  select(-Distance)
```

```{r}
# Filter ASV table only to EED taxa
EED_asv = EED_asv_full %>% 
  select(Sample, one_of(EED_taxonomy$ID))
```


```{r}
# Contribution of exact and potential EED ASVs to total reads
# Total reads per sample
total_reads = EED_asv_full %>% 
  column_to_rownames("Sample") %>% 
  rowSums()

# Exact and potential matches
EED_all_reads = EED_asv %>% 
  column_to_rownames("Sample") %>% 
  rowSums()

reads_sample = data.frame(total_reads, EED_all_reads) %>% 
  rownames_to_column("Sample") %>% 
  mutate(EED_all_per = round((EED_all_reads/total_reads)*100, 2))

reads_sample$EED_all_per %>% 
  summary()

reads_sample %>% 
  filter(EED_all_per < 95) %>% 
  arrange(Sample)
```


## Collapse abundances by taxa
```{r}
# Combine the reads of ASVs with the same taxonomic
EED_taxonomy_collapse = EED_taxonomy %>% 
  select(-c(md5, Seq)) %>% 
  distinct() %>% 
  arrange(Phylum, Class)

EED_asv_long = EED_asv %>% 
  pivot_longer(cols = -Sample, names_to = "ID", values_to = "Reads") %>% 
  left_join(EED_taxonomy_collapse, by = "ID")
  
EED_asv_colapsed = EED_asv_long %>% 
  group_by(Sample, Name) %>% 
  summarize(Sum_reads = sum(Reads)) %>% 
  pivot_wider(id_cols = Sample, names_from = Name, values_from = Sum_reads) %>% 
  ungroup()

```


# Write files
```{r}
# TODO Change the file names if needed
Check_point("Are the file names adjusted?")

# Collapsed abundances
write_tsv(EED_asv_colapsed, 
          file.path(output_dir, "Collapsed_EED.tsv"))

```

# Number of detected species
```{r}
retain_sample = reads_sample %>% 
  filter(EED_all_per > 90) %>% 
  pull(Sample)

EED_asv_colapsed %>% 
  filter(Sample %in% retain_sample) %>% 
  column_to_rownames("Sample") %>% 
  vegan::specnumber() %>% 
  summary()
```


# Barplot of Taxa abundance
```{r fig.height=9, fig.width=6}
EED_pal = c(tableau_color_pal('Tableau 20')(20), stata_pal("s1color")(3))

EED_long_colapsed = EED_asv_colapsed %>% 
  column_to_rownames("Sample") %>% 
  vegan::decostand(method = "total", MARGIN = 1) %>% 
  rownames_to_column("Sample") %>% 
  pivot_longer(cols = -Sample, names_to = "Taxon", values_to = "Relabund")
  
```

```{r fig.height=7, fig.width=12}
retain_sample = reads_sample %>% 
  filter(EED_all_per > 90) %>% 
  pull(Sample)

EED_barplot = EED_long_colapsed %>% 
  filter(Sample %in% retain_sample) %>% 
  ggplot(aes(x = Sample, y = Relabund*100, fill = Taxon)) +
    geom_col(width = 0.5) +
    theme_light() +
    scale_fill_manual(values = EED_pal) +
    theme(text = element_text(size=10), 
          axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(x = "Sample", y = "Relative abundance (%)") +
    guides(fill = guide_legend(ncol = 1)) +
    theme(legend.position="right", 
          text = element_text(size = 10), 
          legend.key.size = unit(0.5, "cm")) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, NA))

EED_barplot



# TODO Change the file names if needed
Check_point("Are the file names adjusted?")

cowplot::save_plot(filename = file.path(fig_dir, "EED_barplot.svg"), EED_barplot, base_height = 7, base_width = 7)
```

# Session Info
```{r}
sessionInfo()
```